### PR TITLE
[79] Implement actor apply jobs UI

### PR DESCRIPTION
**Pull request by fainrada**
## What did you do

image
- Add actor apply page, actor is able to apply job by choosing a resume with dropdown
- link apply button in JobCardFooter to actor apply page
- Add jobId as number type in FooterProps type.ts
 
## Demo

- https://dev.modela.miello.dev/job
- Login with actor
- Choose a job to apply
- Choose resume and click "apply"
- If you already apply to this job, an error snackbar will popup with text "คุณได้สมัครงานนี้ไปแล้ว"
- Else, a success snack bar will popup with text "การสมัครเสร็จสิ้น" and post data to backend

## Limitation

- no unit test

### CODE REVIEW

File: apps/web/src/modules/job/apply/components/SelectResume/index.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import Box from '@mui/material/Box'
+import FormControl from '@mui/material/FormControl'
+import MenuItem from '@mui/material/MenuItem'
+import TextField from 'common/components/TextField'
+import * as React from 'react'
+
+import { ApplyProps } from './types'
+
+export default function SelectResume(props: ApplyProps) {
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/components/SelectResume/index.tsx:11]: you can use `const {id, setId, resumes} = props` 
then use `props.setId(Number(event.target.value))` `value = {id}` and `resumes.map`
for cleaner code

File: apps/web/src/modules/job/apply/components/SelectResume/index.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { ResumeDto } from '@modela/dtos'
+import Box from '@mui/material/Box'
+import FormControl from '@mui/material/FormControl'
+import MenuItem from '@mui/material/MenuItem'
+import TextField from 'common/components/TextField'
+import * as React from 'react'
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/components/SelectResume/index.tsx:6]:
```suggestion
import React from 'react'
```

File: apps/web/src/modules/job/apply/components/SelectResume/index.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          select
+          value={props.Id}
+          label="เลือก resume"
+          onChange={handleChange}
+        >
+          {props.resumes.map((resume: ResumeDto) => (
+            <MenuItem key={resume.resumeId} value={resume.resumeId}>
+              {resume.name}
+            </MenuItem>
+          ))}
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/components/SelectResume/index.tsx:28]: same as above
```suggestion
          {resumes.map(({resumeId, name}: ResumeDto) => (
            <MenuItem key={resumeId} value={resumeId}>
              {name}
            </MenuItem>
          ))}
```

File: apps/web/src/modules/job/apply/components/SelectResume/types.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { ResumeDto } from '@modela/dtos'
+
+export interface ApplyProps {
+  resumes: ResumeDto[]
+  setId: (Id: number) => void
+  Id: number | undefined
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/components/SelectResume/types.ts:6]:
```suggestion
  id: number | undefined
```

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+const useResume = () => {
+  const router = useRouter()
+  const { jobId } = router.query
+  const { handleError } = useErrorHandler()
+  const [loading, setLoading] = useState(false)
+  const [jobTitle, setJobTitle] = useState('')
+  const { displaySnackbar } = useSnackbar()
+
+  const [Id, setId] = React.useState<number | undefined>()
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:17]:
```suggestion
  const [id, setId] = React.useState<number | undefined>()
```

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+  const handleSuccess = useCallback(async () => {
+    setLoading(true)
+    try {
+      const resumeIdBody: ResumeIdDto = {
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:22]: resumeId should be checked before POST to backend (we should not send bad request data)

File: apps/web/src/modules/job/apply/pages/index.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import withGuard from 'common/hoc/withGuard'
+
+import SelectResume from '../components/SelectResume'
+import useResume from './hooks'
+import { CardContainer, RootContainer } from './styled'
+
+const JobApplyPage = () => {
+  const { jobId, jobTitle, handleSuccess, loading, resumes, Id, setId } =
+    useResume()
+  if (loading) return <CircularProgress />
```
Comment (PKhing) [inline apps/web/src/modules/job/apply/pages/index.tsx:18]: My bad I think that this is loading for fetching data, but actually, it is loading for submitting data.
We should move Circular progress to the submit button (see example in `modules/login/index.tsx`)

File: apps/web/src/modules/job/list/components/JobCardContainer/components/JobCard/index.tsx [hunk 8]
```diff
@@ -59,6 +59,7 @@ export default function JobCard(prop: JobCardProps) {
           wage={wage}
           actorCount={actorCount}
           status={status}
+          jobId={prop.jobId}
```
Comment (PKhing) [inline apps/web/src/modules/job/list/components/JobCardContainer/components/JobCard/index.tsx:62]:
```suggestion
          jobId={jobId}
```

**Review by PKhing — CHANGES_REQUESTED** [review]

File: apps/web/src/modules/job/apply/components/SelectResume/types.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { ResumeDto } from '@modela/dtos'
+
+export interface ApplyProps {
+  resumes: ResumeDto[]
+  setId: (Id: number) => void
```
Comment (miello) [inline apps/web/src/modules/job/apply/components/SelectResume/types.ts:5]:
```suggestion
  setId: (id: number) => void
```

File: apps/web/src/modules/job/apply/pages/index.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+          <Link
+            variant="body1"
+            color="primary"
+            sx={{ cursor: 'pointer', textDecoration: 'none' }}
+            href={`/job/${jobId}`}
+          >
+            {jobTitle}
+          </Link>
+        </Typography>
+        <SelectResume resumes={resumes} Id={Id} setId={setId}></SelectResume>
```
Comment (miello) [inline apps/web/src/modules/job/apply/pages/index.tsx:38]: This should fix along with `Id`
```suggestion
        <SelectResume resumes={resumes} id={id} setId={setId}></SelectResume>
```

**Review by miello — CHANGES_REQUESTED** [review]

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import { GetJobDto, GetResumesDto, ResumeDto, ResumeIdDto } from '@modela/dtos'
+import { useSnackbar } from 'common/context/SnackbarContext'
+import { useErrorHandler } from 'common/hooks/useErrorHandler'
+import { apiClient } from 'common/utils/api'
+import { useRouter } from 'next/router'
+import React from 'react'
+import { useCallback, useEffect, useState } from 'react'
```
Comment (miello) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:7]:
```suggestion
import { useCallback, useEffect, useState } from 'react'
```

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+const useResume = () => {
+  const router = useRouter()
+  const { jobId } = router.query
+  const { handleError } = useErrorHandler()
+  const [loading, setLoading] = useState(false)
+  const [jobTitle, setJobTitle] = useState('')
+  const { displaySnackbar } = useSnackbar()
+
+  const [id, setId] = React.useState<number | undefined>()
```
Comment (miello) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:17]:
```suggestion
  const [id, setId] = useState<number | undefined>()
```

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      if (id) {
+        const resumeIdBody: ResumeIdDto = {
+          resumeId: id,
+        }
+        await apiClient.post('jobs/' + jobId + '/apply', resumeIdBody)
+        displaySnackbar('การสมัครเสร็จสิ้น', 'success')
+        router.push('/job', undefined, { shallow: true })
+      } else {
+        displaySnackbar('กรุณาเลือกเรซูเม่', 'error')
+      }
```
Comment (miello) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:31]: (Discussion)
Should we change handle missing resumeId from `displaySnackbar` to assign message to `helperText` in TextField instead ?

Comment (miello) [reply apps/web/src/modules/job/apply/pages/hooks/index.ts:31]: Please see other page such as login, register for reference

File: apps/web/src/modules/job/apply/pages/hooks/index.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+      handleError(err, {
+        409: 'คุณได้สมัครงานนี้ไปแล้ว',
+        400: 'กรุญาเลือกเรซูเม่',
+      })
+    } finally {
+      setLoading(false)
+    }
+  }, [id, displaySnackbar, handleError, jobId, router])
+
+  const [resumes, setResumes] = React.useState<ResumeDto[]>([])
```
Comment (miello) [inline apps/web/src/modules/job/apply/pages/hooks/index.ts:42]:
```suggestion
  const [resumes, setResumes] = useState<ResumeDto[]>([])
```

**Review by miello — CHANGES_REQUESTED** [review]

**Review by PKhing — APPROVED** [review]

**Review by miello — APPROVED** [review]

**Pull request merged**